content/concepts/did-ddo.md

```diff
@@ -1,173 +1,551 @@
 
-If a DID is the index key in a key-value pair, then the DID Document is the value to which the index key points.
-The combination of a DID and its associated DID Document forms the root record for a decentralized identifier.
+An _asset_ should have a DID and DDO. The DDO should include [metadata](#metadata) about the asset, and define access in at least one [service](#services). The DDO can only be modified by _owners_ or _delegated users_.
 
-DIDs and DDOs follow [this specification](https://w3c-ccg.github.io/did-spec/) defined by the World Wide Web Consurtium (W3C).
+A metadata cache like _Aquarius_ can help in reading and searching through encrypted DDO data from the chain.
 
-## Rules for DIDs & DDOs in Ocean
+## Publishing & Retrieving DDOs
 
-- An _asset_ in Ocean represents a downloadable file, compute service, or similar. Each asset is a _resource_ under control of a _publisher_. The Ocean network itself does _not_ store the actual resource (e.g. files).
-- An asset should have a DID and DDO. The DDO should include metadata about the asset.
-- The DDO can only can be modified by _owners_ or _delegated users_.
-- There _must_ be at least one client library acting as _resolver_, to get a DDO from a DID.
-- The DDO is stored on-chain. It's stored in in plaintext, with two exceptions: (1) the field for resource-access url is encrypted (2) the whole DDO may be encrypted, if the publisher is willing to lose 100% of discoverability.
-- A metadata cache like Aquarius can help in reading and writing DDO data from the chain.
+The DDO is stored on-chain as part of the NFT contract and it is stored encrypted using the private key of the _Provider_. To resolve it, a metadata cache like _Aquarius_ must query the provider to decrypt the DDO.
 
-## DID Structure
+Here is the complete flow:
+
+![DDO_flow](images/ddo-flow.png)
+
+```text
+title DDO flow
+
+User(Ocean library) -> User(Ocean library): Prepare DDO
+User(Ocean library) -> Provider: encrypt DDO
+Provider -> User(Ocean library): encryptedDDO
+User(Ocean library) -> ERC721 contract: publish encryptedDDO
+Aquarius <-> ERC721 contract: monitors ERC721 contract and gets MetdadataCreated Event (contains encryptedDDO)
+Aquarius -> ERC721 contract: calls getMetaData()
+Aquarius -> Provider: decrypt encryptedDDO, signed request using Aquarius's private key
+Provider -> ERC721 contract: checks state using getMetaData()
+Provider -> Provider: depending on metadataState (expired,retired) and aquarius address, validates the request
+Provider -> Aquarius: DDO
+Aquarius -> Aquarius : validate DDO
+Aquarius -> Aquarius : cache DDO
+Aquarius -> Aquarius : enhance cached DDO in response with additional infos like `events` & `stats`
+```
+
+## DID
 
 In Ocean, a DID is a string that looks like:
 
 ```text
 did:op:0ebed8226ada17fde24b6bf2b95d27f8f05fcce09139ff5cec31f6d81a7cd2ea
 ```
 
+where
+
+```text
+0ebed8226ada17fde24b6bf2b95d27f8f05fcce09139ff5cec31f6d81a7cd2ea = sha256(ERC721 contract address + chainId)
```

Comment:
Would it make sense dropping all the custom OCEAN DID logic and just use cross-chain identifies of: https://github.com/ChainAgnostic/CAIPs ?

Seems anyways, they're doing exactly what's needed here.

Reply:
you always know the good bleeding edge stuff. Guess https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-19.md would apply. 

Honestly, personally, I would love to simply stick to some external standard as we keep changing the DID in every version of the protocol which kind of defeats the purpose of a DID in the long term. Using a standard solving the problems we try to solve would be better insurance against not having to change it again in v5. 

Just not sure about actual implications and existing implementations as the `did:op:` prefix is registered in some DID registry. With no actual practical real use I would assume. But Ocean is also unique, here specifically having the dataset represented as an NFT so using the NFT address in some form still makes more sense.

But let's do hypothetic play, guess we would then have something like this for an asset own ETH Mainnet:

```text
eip155:1/op:nft_address

or 

eip155:1/ocean:nft_address

# assuming this won't work based on CAIP-19
eip155:1/did:op:nft_address
```

Which to me looks like interesting routing problems when used as is in a URL like `/asset/DID`, the slash really is a problem there as it would need to be made escaped all the time, eliminating a use case where user types it in. Other than that, if it helps us to stop changing the DID format, would be all for it

Reply:
> Which to me looks like interesting routing problems when used as is in a URL like `/asset/DID`, the slash really is a problem there as it would need to be made escaped all the time, eliminating a use case where user types it in.

That's a really great point and I'm amused to see that CAIP went with a slash symbol. Bad choice IMO.



> Honestly, personally, I would love to simply stick to some external standard as we keep changing the DID in every version of the protocol which kind of defeats the purpose of a DID in the long term.

Totally agreed. I commented here as I also feel slightly too lazy to refactor my whole codebase to didv4. For an external user like rugpullindex.com there's an added benefit in using a global standard, which is that the DIDOPv3 really was just a pointer to an ERC20 on ETH mainnet initially (chainID:1 was implicitly assumed). So had I used this as my baseline assumption when architecting RPI, then it'd now be easier managing other chainID:1 type assets too. My code base is biased by DIDv3. Long term, I wanna go towards an identifier that is as long-lasting and generic as possible. CAIP looks interesting but I'm still exploring too.

Regarding your question on usage. IMO, a better fit for OP would e.g. be caip-22:

```
eip155:1/erc721:0x06012c8cf97BEaD5deAe237070F9587f8E7A266d
```

where the OCEAN NFT is simply addressed using chain id and where it's obvious that it's based on an NFT/ERC721 standard. Or for a data token ERC20 caip-21

```
eip155:1/erc20:0x6b175474e89094c44da98b954eedeac495271d0f
```

Now, I understand that your point of using the DID is also to identify the higher-level concept in OCEAN that represents "a data set", which is actually a couple of separate things:

- a JSON object called DDO
- an identifier, the DIDv3
- 1-n AMM pools where ERC20 tokens are traded
- 1 ERC20 token
- A data publisher represented by 1 Ethereum-style hex address `0x...`

For "the data set", I'm not sure if CAIP has a good standard yet as I believe conceptually the space isn't evolved far enough yet. What is a "data set" really? Is a higher-level concept required? Or can a data set be an NFT/ERC721?

From the docs I commented on, I can see that DIDv4 uses the data token address on a chain as the proxy for "the data set identifier." I'd suggest:

- Addressing the data token using e.g. CAIP-22
- Addressing a data NFT, using e.g. CAIP-21

IMO, this can already be a big leap forward as it'd enable OP to become layer3 (in the evm-space) as all identifiers contain e.g. a chainID based on EIP155.

As for the "the data set identifier". I'm not sure. It's difficult to wrap my head around that ATM.

Reply:
I would not do this.  Never, never we should rely on the fact that the asset id has a meaning. (We did that mistake in v3).
What is the advantage?  What about later, when we will switch from erc721 to etc21433 ?  Change it all again ?

a DID is just a unique id that refers to an asset.  An asset without a DDO is useless.

Reply:
Would not do what? What do you mean by "has a meaning"?

>  What about later, when we will switch from erc721 to etc21433 ? Change it all again ?

When you switch from ERC721 to ETC21433, you address a new digital artifact and so you adjust the identifier.

Reply:
Besides @alexcos20 I doubt that speculating about any technical vision is helpful on this issue. Whether or not you'll switch "from erc721 to etc21433" isn't helpful to ask.

Both you and I know exactly nothing about the future requirements of an OCEAN identifier as we both can't forecast the future.

My argument here is mainly concerned with correctly addressing digital artifacts of OCEAN V4, and if there's an ERC20 and an ERC721 to address on multiple chains, then I think CAIP is a great choice.

Reply:
if we drop "did:op:XX", means that we are going to drop the W3C standard.  and we are using that not for "did:op:", but also for the DDO definition.  switching to CAIP will give us a hybrid schema (CAIP for id, W3C for DDO), which is cumbersome.

Again, what is the benefit? Just to get the ERC721 address from the ID (because we cannot call it did anymore)?

We have 2 scenarios:
 - you have the did, resolve it and get the DDO, then use DDO.nft.address
 - you have the ERC721 address.. do sha256(erc721Address+chainID) and get the DID